app/src/mainwindow2.cpp

```diff
@@ -1380,10 +1380,19 @@ void MainWindow2::makeConnections(Editor* editor, ScribbleArea* scribbleArea)
 {
     connect(editor->tools(), &ToolManager::toolChanged, scribbleArea, &ScribbleArea::setCurrentTool);
     connect(editor->tools(), &ToolManager::toolPropertyChanged, scribbleArea, &ScribbleArea::updateToolCursor);
-    connect(editor->layers(), &LayerManager::currentLayerChanged, scribbleArea, &ScribbleArea::updateAllFrames);
+    connect(editor->layers(), &LayerManager::currentLayerChanged, [editor, scribbleArea]
+    {
+        // Changing the current layer will only change the frame (as viewed by the user) under the following circumstances
+        if(editor->preference()->isOn(SETTING::PREV_ONION) ||
+           editor->preference()->isOn(SETTING::NEXT_ONION) ||
+           scribbleArea->getLayerVisibility() != LayerVisibility::ALL)
+        {
```

Comment:
Is this a necessary check? isn't it better to just update no matter what reason the layer changed?

Reply:
When we're talking about necessary, it's only *necessary* to update the frames when the current layer changes if onion skinning or relative layer visibility is enabled. `updateAllFrames` invalidates the entire cache, so it should be treated as a very costly operation that is only done when it is absolutely needed. A couple checks here could prevent hundreds of frames from being re-rendered.

It might be a bit better to always update all frames to avoid visual artifacts, particularly when some new feature is added which also needs to be updated when the current layer changes. However if we follow that logic to it's natural conclusion, it would be best to get rid of the cache altogether to prevent that issue under any circumstance, which I don't agree with.

Reply:
But clearing the entire cache may also be overkill, even if you enable onion skin, you'll likely never have more than 5 frames skinned. Maybe it would be better to clear only the amount of frames that has been drawn. Sure I can see that it's an expensive operation if you have hundreds of frames but I don't see why you would invalidate all those frames either, if you know the onion skin only has been drawn on 3-5 frames.

There's no natural conclusion, of course the cache is important and should stay but checking various cases, might lead to other issues in the future, especially if we are going to implement features that might require similar handling.

I'm just saying that depending on the situation, it might be okay to not have a check, if the opposite is that we get weird artifacts in some edge case we don't know.

Alternatively I think we should have this logic in a function, so one is less likely to miss it, since it's quite performance relevant.

Reply:
I don't really understand what you want here. You think clearing the cache is overkill, but then you want to remove the conditional and always clear the cache?

I should clarify here what exactly I mean when I say cache in the context of this particular discussion. What we are referring to here is the QPixmapCache in ScribbleArea. This holds *rendered* frames, which includes the onion skin. The onion skin is only for the current layer, so when you switch layers it changes. If for example you:
1. View frame 7 with the first layer active and onion skinning enabled
2. Go to frame 100
3. Switch to the second layer

Now the cached frame 7 has the onion skin for layer 1, not layer 2, and thus needs to be redrawn. Same with every other frame in the cache.

I will move it to its own function if you really want. I can't think of any other situation where you would use it, but maybe it would make it a bit easier to find.

Reply:
> I assume that you mean it prevents hundreds of frames from being re-rendered, because the entire cache is being cleared, that's why I mention that instead of clearing the whole cache, why not clear only the affected frames, then it won't be as costly.

I completely agree with that sentiment. As I tried to explain in the previous message, all of the frames are affected when changing the active layer with onion skinning enabled (and the same logic applies to relative layer visibility), and none of the frames are affected otherwise, hence the check.

Reply:
Ah... aaah right, yes okay I get what you're saying now :P Alright then what I'm suggesting doesn't make much sense.